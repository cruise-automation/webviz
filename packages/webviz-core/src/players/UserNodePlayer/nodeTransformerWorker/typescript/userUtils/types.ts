export type RGBA = {
  // all values are scaled between 0-1 instead of 0-255
  r: number;
  g: number;
  b: number;
  a: number; // opacity -- typically you should set this to 1.
};

export type Header = {
  frame_id: string;
  stamp: Time;
  seq: number;
};

export type Point = {
  x: number;
  y: number;
  z: number;
};

export type Time = {
  sec: number;
  nsec: number;
};

export type Translation = {
  x: number;
  y: number;
  z: number;
};

export type Rotation = {
  x: number;
  y: number;
  z: number;
  w: number;
};

export type Pose = {
  position: Point;
  orientation: Quaternion;
};

export type Quaternion = {
  x: number;
  y: number;
  z: number;
  w: number;
};

export type Transform = {
  header: Header;
  child_frame_id: string;
  transform: {
    translation: Translation;
    rotation: Rotation;
  };
};

export type PointStyle =
  | "circle"
  | "cross"
  | "crossRot"
  | "dash"
  | "line"
  | "rect"
  | "rectRounded"
  | "rectRot"
  | "star"
  | "triangle";

export type TwoDimensionalPlotLine = {
  order?: number;
  label: string;
  backgroundColor?: string;
  borderColor?: string;
  borderDash?: number[];
  borderWidth?: number;
  pointBackgroundColor?: string;
  pointBorderColor?: string;
  pointBorderWidth?: number;
  pointRadius?: number;
  pointStyle?: PointStyle;
  lineTension?: number;
  data: { x: number; y: number }[];
};
/*
 * Use this type to plot values in the 2D Panel. If you need to change this
 * type, please ensure that is kept in sync with the TwoDimensionalPlot
 * counterpart.
 */
export type TwoDimensionalPlot = {
  lines: TwoDimensionalPlotLine[];
  points?: TwoDimensionalPlotLine[];
  polygons?: TwoDimensionalPlotLine[];
  title?: string;
  yAxisLabel?: string;
  xAxisLabel?: string;
  gridColor?: string;
};
