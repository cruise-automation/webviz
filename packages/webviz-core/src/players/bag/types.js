// @flow
//
//  Copyright (c) 2018-present, GM Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { type Time } from "rosbag";

export type InitializeMessage = {
  bagPath: File | string,
};

export type MessageRequest = {
  start: Time,
  end: Time,
  topics: string[],
};

export type RawMessage = {
  topic: string,
  buffer: ArrayBuffer,
  timestamp: Time,
};
