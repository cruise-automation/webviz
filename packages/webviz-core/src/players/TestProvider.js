// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { type Time } from "rosbag";

import type {
  ExtensionPoint,
  InitializationResult,
  DataProvider,
  GetMessagesResult,
  GetMessagesTopics,
} from "webviz-core/src/dataProviders/types";
import type { Topic } from "webviz-core/src/players/types";
import type { RosDatatypes } from "webviz-core/src/types/RosDatatypes";

export const defaultStart = { sec: 10, nsec: 0 };
export const defaultEnd = { sec: 100, nsec: 0 };
const datatypes: RosDatatypes = {
  fooBar: {
    fields: [
      {
        name: "val",
        type: "number",
      },
    ],
  },
  baz: {
    fields: [
      {
        name: "val",
        type: "number",
      },
    ],
  },
};
const defaultTopics: Topic[] = [{ name: "/foo/bar", datatype: "fooBar" }, { name: "/baz", datatype: "baz" }];
type GetMessages = (start: Time, end: Time, topics: GetMessagesTopics) => Promise<GetMessagesResult>;

export default class TestProvider implements DataProvider {
  _start: Time;
  _end: Time;
  _topics: Topic[];
  _datatypes: RosDatatypes;
  extensionPoint: ExtensionPoint;
  closed: boolean = false;

  constructor({ getMessages, topics }: { getMessages?: GetMessages, topics?: Topic[] } = {}) {
    this._start = defaultStart;
    this._end = defaultEnd;
    this._topics = topics ?? defaultTopics;
    this._datatypes = datatypes;
    if (getMessages) {
      this.getMessages = getMessages;
    }
  }

  initialize(extensionPoint: ExtensionPoint): Promise<InitializationResult> {
    this.extensionPoint = extensionPoint;
    return Promise.resolve({
      start: this._start,
      end: this._end,
      topics: this._topics,
      providesParsedMessages: true,
      messageDefinitions: {
        type: "parsed",
        datatypes: this._datatypes,
        messageDefinitionsByTopic: {},
        parsedMessageDefinitionsByTopic: {},
      },
    });
  }

  getMessages: GetMessages = (_start: Time, _end: Time, _topics: GetMessagesTopics): Promise<GetMessagesResult> => {
    throw new Error("not implemented");
  };

  close(): Promise<void> {
    this.closed = true;
    return Promise.resolve();
  }
}
