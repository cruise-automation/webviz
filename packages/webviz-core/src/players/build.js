// @flow
//
//  Copyright (c) 2018-present, GM Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { playerDisconnected, getPlayer, getPipeline, setPlayerNull } from "webviz-core/src/actions/player";
import { BagDataProvider } from "webviz-core/src/players/bag/BagDataProvider";
import CombinedDataProvider from "webviz-core/src/players/CombinedDataProvider";
import RandomAccessPlayer from "webviz-core/src/players/RandomAccessPlayer";
import ReadAheadDataProvider from "webviz-core/src/players/ReadAheadDataProvider";
import type { Player } from "webviz-core/src/types/players";
import type { Dispatch, GetState } from "webviz-core/src/types/Store";
import { SECOND_BAG_PREFIX } from "webviz-core/src/util/globalConstants";
import reportError from "webviz-core/src/util/reportError";

export type PlayerInput = string | File | Player;

let lastBagPayload: ?File;

function clearPlayer(dispatch: Dispatch) {
  const player = getPlayer();
  if (player) {
    player.close();
  }
  setPlayerNull();
  dispatch(playerDisconnected());
}

export const loadBag = (files: FileList | File[], addBag: boolean) => async (
  dispatch: Dispatch,
  getState: GetState
): Promise<void> => {
  if (getPlayer()) {
    clearPlayer(dispatch);
  }
  const file = files[0];
  let bagProvider;
  if (lastBagPayload && addBag) {
    const bagProvider1 = new BagDataProvider(lastBagPayload);
    const bagProvider2 = new BagDataProvider(file);
    bagProvider = new CombinedDataProvider([
      { provider: bagProvider1 },
      { provider: bagProvider2, prefix: SECOND_BAG_PREFIX },
    ]);
    lastBagPayload = undefined;
  } else {
    bagProvider = new BagDataProvider(file);
    lastBagPayload = file;
  }
  const provider = new ReadAheadDataProvider(bagProvider);
  const player = new RandomAccessPlayer(provider, undefined, true);

  try {
    await getPipeline().initialize(dispatch, player);
  } catch (e) {
    reportError("Player failed to initialize", e);
    return clearPlayer(dispatch);
  }
};
