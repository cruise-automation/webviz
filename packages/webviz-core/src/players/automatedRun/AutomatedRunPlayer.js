// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { partition } from "lodash";
import Queue from "promise-queue";
import { type Time, TimeUtil } from "rosbag";
import uuid from "uuid";

import type { DataProvider, DataProviderMetadata, InitializationResult } from "webviz-core/src/dataProviders/types";
import { BUFFER_DURATION_SECS } from "webviz-core/src/players/OrderedStampPlayer";
import { SEEK_BACK_NANOSECONDS } from "webviz-core/src/players/RandomAccessPlayer";
import type {
  AdvertisePayload,
  BobjectMessage,
  Message,
  Player,
  PlayerState,
  Progress,
  PublishPayload,
  SubscribePayload,
  Topic,
} from "webviz-core/src/players/types";
import { USER_ERROR_PREFIX } from "webviz-core/src/util/globalConstants";
import { getSanitizedTopics } from "webviz-core/src/util/selectors";
import sendNotification, {
  type NotificationType,
  type NotificationSeverity,
  type DetailsType,
  detailsToString,
  setNotificationHandler,
} from "webviz-core/src/util/sendNotification";
import {
  fromSec,
  getSeekTimeFromSpec,
  clampTime,
  subtractTimes,
  toMillis,
  fromMillis,
  getSeekToTime,
  type TimestampMethod,
} from "webviz-core/src/util/time";

export type VideoMetadata = {|
  startTimeMs: number, // time of first message in millis (inclusive)
  endTimeMs: number, // time of last message in millis (inclusive)
  msPerFrame: number, // millis per frame
|};

export interface AutomatedRunClient {
  speed: number;
  msPerFrame: number;
  durationMs?: number | typeof undefined;
  workerIndex?: number;
  workerTotal?: number;
  shouldLoadDataBeforePlaying: boolean;
  onError(any): Promise<void>;
  start({ bagLengthMs: number }): void;
  markTotalFrameStart(): void;
  markTotalFrameEnd(): void;
  markFrameRenderStart(): void;
  markFrameRenderEnd(): number;
  markPreloadStart(): void;
  markPreloadEnd(): number;
  onFrameFinished(): Promise<void>;
  finish(VideoMetadata): any;
}

export const AUTOMATED_RUN_START_DELAY = process.env.NODE_ENV === "test" ? 10 : 2000;
const NO_WARNINGS = Object.freeze({});

function formatSeconds(sec: number): string {
  const date = new Date(0);
  date.setSeconds(sec);
  return date.toISOString().substr(11, 8);
}

export default class AutomatedRunPlayer implements Player {
  static className = "AutomatedRunPlayer";
  _isPlaying: boolean = false;
  _provider: DataProvider;
  _startTime: Time;
  _providerResult: InitializationResult;
  _providerTopics: Topic[] = [];
  _progress: Progress;
  _bobjectTopics: Set<string> = new Set();
  _parsedTopics: Set<string> = new Set();
  _listener: (PlayerState) => Promise<void>;
  _initializeTimeout: TimeoutID;
  _initialized: boolean = false;
  _id: string = uuid.v4();
  _speed: number;
  _msPerFrame: number;
  _client: AutomatedRunClient;
  _error: ?Error;
  _waitToReportErrorPromise: ?Promise<void>;
  _startCalled: boolean = false;
  _receivedBytes: number = 0;
  _globalMessageOrder: TimestampMethod = "receiveTime";
  // Calls to this._listener must not happen concurrently, and we want them to happen
  // deterministically so we put them in a FIFO queue.
  _emitStateQueue: Queue = new Queue(1);

  constructor(provider: DataProvider, client: AutomatedRunClient) {
    this._provider = provider;
    this._speed = client.speed;
    this._msPerFrame = client.msPerFrame;
    this._client = client;
    // Report errors from sendNotification and those thrown on the window object to the client.
    setNotificationHandler(
      (message: string, details: DetailsType, type: NotificationType, severity: NotificationSeverity) => {
        if (severity === "warn") {
          // We can ignore warnings in automated runs
          return;
        }
        let error;
        if (type === "user") {
          error = new Error(`${USER_ERROR_PREFIX} ${message} // ${detailsToString(details)}`);
        } else if (type === "app") {
          error = new Error(`[WEBVIZ APPLICATION ERROR] ${detailsToString(details)}`);
        } else {
          (type: void);
          error = new Error(`Unknown error type! ${type} // ${detailsToString(details)}`);
        }
        this._error = error;
        this._waitToReportErrorPromise = client.onError(error);
      }
    );
    window.addEventListener("error", (e: Error) => {
      // This can happen when ResizeObserver can't resolve its callbacks fast enough, but we can ignore it.
      // See https://stackoverflow.com/questions/49384120/resizeobserver-loop-limit-exceeded
      if (e.message.includes("ResizeObserver loop limit exceeded")) {
        return;
      }
      this._error = e;
      this._waitToReportErrorPromise = client.onError(e);
    });
  }

  _end(): Time {
    // Play "past the end" when in header-stamp mode so the OrderedStampPlayer goes up to the end.
    return this._globalMessageOrder === "receiveTime"
      ? this._providerResult.end
      : TimeUtil.add(this._providerResult.end, fromSec(BUFFER_DURATION_SECS));
  }

  async _getMessages(
    start: Time,
    end: Time
  ): Promise<{ parsedMessages: $ReadOnlyArray<Message>, bobjects: $ReadOnlyArray<BobjectMessage> }> {
    const parsedTopics = getSanitizedTopics(this._parsedTopics, this._providerTopics);
    const bobjectTopics = getSanitizedTopics(this._bobjectTopics, this._providerTopics);
    if (parsedTopics.length === 0 && bobjectTopics.length === 0) {
      return { parsedMessages: [], bobjects: [] };
    }
    // Make sure to clamp the start/end times to the _providerResult or getMessages will throw errors
    // TODO: Find a better way to share this logic with RandomAccessPlayer.js
    start = clampTime(start, this._providerResult.start, this._providerResult.end);
    end = clampTime(end, this._providerResult.start, this._providerResult.end);
    const messages = await this._provider.getMessages(start, end, {
      parsedMessages: parsedTopics,
      bobjects: bobjectTopics,
    });
    const { parsedMessages, rosBinaryMessages, bobjects } = messages;
    if (rosBinaryMessages?.length || bobjects == null || parsedMessages == null) {
      const messageTypes = Object.keys(messages)
        .filter((kind) => messages[kind]?.length)
        .join(",");
      throw new Error(`Invalid message types: ${messageTypes}`);
    }

    const filterMessages = (msgs) =>
      msgs.map((message) => {
        const topic: ?Topic = this._providerTopics.find((t) => t.name === message.topic);
        if (!topic) {
          throw new Error(`Could not find topic for message ${message.topic}`);
        }

        if (!topic.datatype) {
          throw new Error(`Missing datatype for topic: ${message.topic}`);
        }
        return {
          topic: message.topic,
          receiveTime: message.receiveTime,
          message: message.message,
        };
      });
    return { parsedMessages: filterMessages(parsedMessages), bobjects: filterMessages(bobjects) };
  }

  _emitState(
    messages: $ReadOnlyArray<Message>,
    bobjects: $ReadOnlyArray<BobjectMessage>,
    currentTime: Time
  ): Promise<void> {
    return this._emitStateQueue.add(async () => {
      if (!this._listener) {
        return;
      }
      const endTime = this._end();
      const initializationResult = this._providerResult;
      if (!initializationResult.messageDefinitions || initializationResult.messageDefinitions.type === "raw") {
        throw new Error("AutomatedRunPlayer requires parsed message definitions");
      }
      return this._listener({
        isPresent: true,
        showSpinner: false,
        showInitializing: false,
        progress: this._progress,
        capabilities: [],
        playerId: this._id,
        activeData: {
          messages,
          bobjects,
          totalBytesReceived: this._receivedBytes,
          currentTime,
          startTime: this._providerResult.start,
          endTime,
          isPlaying: this._isPlaying,
          speed: this._speed,
          messageOrder: "receiveTime",
          lastSeekTime: 0,
          topics: this._providerTopics,
          datatypes: initializationResult.messageDefinitions.datatypes,
          parsedMessageDefinitionsByTopic: initializationResult.messageDefinitions.parsedMessageDefinitionsByTopic,
          playerWarnings: NO_WARNINGS,
        },
      });
    });
  }

  setListener(callback: (PlayerState) => Promise<void>): void {
    this._listener = callback;
  }

  setSubscriptions(subscriptions: SubscribePayload[]): void {
    const [bobjectSubscriptions, parsedSubscriptions] = partition(subscriptions, ({ format }) => format === "bobjects");
    this._bobjectTopics = new Set(bobjectSubscriptions.map(({ topic }) => topic));
    this._parsedTopics = new Set(parsedSubscriptions.map(({ topic }) => topic));

    // Wait with running until we've subscribed to a bunch of topics.
    clearTimeout(this._initializeTimeout);
    this._initializeTimeout = setTimeout(() => this._initialize(), AUTOMATED_RUN_START_DELAY);
  }

  async _initialize() {
    if (this._initialized) {
      return; // Prevent double loads.
    }
    this._initialized = true;

    this._providerResult = await this._provider.initialize({
      progressCallback: (progress: Progress) => {
        this._progress = progress;
        this._onUpdateProgress();
      },
      reportMetadataCallback: (metadata: DataProviderMetadata) => {
        switch (metadata.type) {
          case "updateReconnecting":
            sendNotification(
              "updateReconnecting should never be called here",
              `AutomatedRunPlayer only supports local playback`,
              "app",
              "error"
            );
            break;
          case "average_throughput":
            // Don't need analytics for data provider callbacks in video generation.
            break;
          case "initializationPerformance":
            break;
          case "received_bytes":
            this._receivedBytes += metadata.bytes;
            break;
          case "data_provider_stall":
            break;
          default:
            (metadata.type: empty);
        }
      },
      notifyPlayerManager: async () => {},
    });
    this._providerTopics = this._providerResult.topics.map((t) => ({ ...t, preloadable: true }));
    this._startTime = getSeekTimeFromSpec(getSeekToTime(), this._providerResult.start, this._end());
    await this._start();
  }

  async _start() {
    // Call _getMessages to start data loading and rendering for the first frame.
    const backfillStart = subtractTimes(this._startTime, { sec: 0, nsec: SEEK_BACK_NANOSECONDS });
    const { parsedMessages, bobjects } = await this._getMessages(backfillStart, this._startTime);
    await this._emitState(parsedMessages, bobjects, backfillStart);
    if (!this._startCalled) {
      this._client.markPreloadStart();
    }

    this._startCalled = true;
    this._maybeStartPlayback();
  }

  async _onUpdateProgress() {
    if (this._client.shouldLoadDataBeforePlaying && this._providerResult != null) {
      // Update the view and do preloading calculations. Not necessary if we're already playing.
      this._emitState([], [], this._startTime);
    }
    this._maybeStartPlayback();
  }

  async _maybeStartPlayback() {
    if (this._readyToPlay()) {
      this._run();
    }
  }

  _readyToPlay() {
    if (!this._startCalled || this._providerResult == null) {
      return false;
    }
    if (!this._client.shouldLoadDataBeforePlaying) {
      return true;
    }
    // If the client has shouldLoadDataBeforePlaying set to true, only start playback once all data has loaded.
    return (
      this._progress &&
      this._progress.fullyLoadedFractionRanges &&
      this._progress.fullyLoadedFractionRanges.length &&
      this._progress.fullyLoadedFractionRanges.every(({ start, end }) => start === 0 && end === 1)
    );
  }

  async _run() {
    if (this._isPlaying) {
      return; // Only run once
    }
    this._isPlaying = true;
    this._client.markPreloadEnd();
    console.log("AutomatedRunPlayer._run()");
    await this._emitState([], [], this._startTime);

    const requestedDurationTime = this._client.durationMs
      ? fromMillis(this._client.durationMs)
      : { sec: Infinity, nsec: 0 };
    const endTime = clampTime(
      TimeUtil.add(this._startTime, requestedDurationTime),
      this._providerResult.start,
      this._end()
    );
    const videoDurationMs = toMillis(subtractTimes(endTime, this._startTime));
    this._client.start({ bagLengthMs: videoDurationMs });

    const startEpoch = Date.now();
    const nsBagTimePerFrame = Math.round(this._msPerFrame * this._speed * 1000000);

    // We split up the frames between the workers,
    // so we need to advance time based on the number of workers
    const workerIndex = this._client.workerIndex ?? 0;
    const workerCount = this._client.workerTotal ?? 1;
    const nsFrameTimePerWorker = nsBagTimePerFrame * workerCount;
    let currentTime = TimeUtil.add(this._startTime, { sec: 0, nsec: nsBagTimePerFrame * workerIndex });

    // Main video recording loop
    while (TimeUtil.isLessThan(currentTime, endTime)) {
      if (this._waitToReportErrorPromise) {
        await this._waitToReportErrorPromise;
      }

      this._client.markTotalFrameStart();

      const internalBackfillDuration = { sec: 0, nsec: SEEK_BACK_NANOSECONDS };
      const backfillStart = clampTime(
        subtractTimes(currentTime, internalBackfillDuration),
        this._startTime,
        this._end()
      );
      const { parsedMessages, bobjects } = await this._getMessages(backfillStart, currentTime);
      this._client.markFrameRenderStart();

      // Wait for the frame render to finish.
      await this._emitState(parsedMessages, bobjects, currentTime);

      this._client.markTotalFrameEnd();
      const frameRenderDurationMs = this._client.markFrameRenderEnd();

      const bagTimeSinceStartMs = toMillis(subtractTimes(currentTime, this._startTime));
      currentTime = clampTime(
        TimeUtil.add(currentTime, { sec: 0, nsec: nsFrameTimePerWorker }),
        this._providerResult.start,
        this._end()
      );

      const percentComplete = bagTimeSinceStartMs / videoDurationMs;
      const msPerPercent = (Date.now() - startEpoch) / percentComplete;
      const estimatedSecondsRemaining = Math.round(((1 - percentComplete) * msPerPercent) / 1000);
      const eta = formatSeconds(Math.min(estimatedSecondsRemaining || 0, 24 * 60 * 60 /* 24 hours */));
      console.log(
        `[${workerIndex}/${workerCount}] Recording ${(percentComplete * 100).toFixed(
          1
        )}% done. ETA: ${eta}. Frame took ${frameRenderDurationMs}ms`
      );
      await this._client.onFrameFinished();
    }

    await this._client.finish({
      startTimeMs: toMillis(this._startTime),
      endTimeMs: toMillis(endTime),
      msPerFrame: this._msPerFrame,
    });
    const totalDuration = (Date.now() - startEpoch) / 1000;
    console.log(`AutomatedRunPlayer finished in ${formatSeconds(totalDuration)}`);
  }

  /* Public API shared functions */

  requestMessages() {}

  setPublishers(_publishers: AdvertisePayload[]) {}

  publish(_payload: PublishPayload) {
    throw new Error(`Unsupported in AutomatedRunPlayer`);
  }

  close() {
    console.warn(`close: Unsupported in AutomatedRunPlayer`);
  }

  startPlayback() {
    console.warn(`startPlayback: Unsupported in AutomatedRunPlayer`);
  }

  pausePlayback() {
    console.warn(`pausePlayback: Unsupported in AutomatedRunPlayer`);
  }

  setPlaybackSpeed(_speed: number, _backfillDuration: ?Time) {
    // This should be passed into the constructor and should not be changed.
  }

  seekPlayback(_time: Time) {
    console.warn(`seekPlayback: Unsupported in AutomatedRunPlayer`);
  }

  requestBackfill() {}
  setGlobalVariables() {
    console.warn(`setGlobalVariables: Unsupported in AutomatedRunPlayer`);
  }
  setMessageOrder(messageOrder: TimestampMethod) {
    this._globalMessageOrder = messageOrder;
  }
}
