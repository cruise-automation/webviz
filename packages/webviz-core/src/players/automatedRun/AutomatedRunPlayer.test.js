// @flow
//
//  Copyright (c) 2018-present, Cruise LLC
//
//  This source code is licensed under the Apache License, Version 2.0,
//  found in the LICENSE file in the root directory of this source tree.
//  You may not use this file except in compliance with the License.

import { TimeUtil } from "rosbag";

import TestProvider, { defaultStart, defaultEnd } from "../TestProvider";
import AutomatedRunPlayer, { type AutomatedRunClient, AUTOMATED_RUN_START_DELAY } from "./AutomatedRunPlayer";
import delay from "webviz-core/shared/delay";
import signal from "webviz-core/shared/signal";
import { type Progress } from "webviz-core/src/players/types";

class TestRunClient implements AutomatedRunClient {
  speed = 1;
  msPerFrame = 20000;
  shouldLoadDataBeforePlaying = false;

  finished = false;
  frameStarted = false;

  constructor({ shouldLoadDataBeforePlaying }: { shouldLoadDataBeforePlaying?: boolean } = {}) {
    this.shouldLoadDataBeforePlaying = shouldLoadDataBeforePlaying || false;
  }

  onError(err) {
    throw err;
  }
  start() {}
  markTotalFrameStart() {
    this.frameStarted = true;
  }
  markTotalFrameEnd() {}
  markFrameRenderStart() {}
  markFrameRenderEnd() {
    return 0;
  }
  markPreloadStart = jest.fn();
  markPreloadEnd = jest.fn();
  async onFrameFinished() {}
  finish() {
    this.finished = true;
  }
}

const getMessagesResult = { parsedMessages: [], rosBinaryMessages: undefined, bobjects: [] };

/* eslint-disable no-underscore-dangle */
describe("AutomatedRunPlayer", () => {
  it("waits to start playing until all frames are loaded when shouldLoadDataBeforePlaying=true", async () => {
    const provider = new TestProvider({ getMessages: async () => getMessagesResult });
    const client = new TestRunClient({ shouldLoadDataBeforePlaying: true });
    const player = new AutomatedRunPlayer(provider, client);
    player.setSubscriptions([{ topic: "/foo/bar", format: "parsedMessages" }]);
    await delay(AUTOMATED_RUN_START_DELAY + 10);
    expect(player._initialized).toEqual(true);
    expect(player._isPlaying).toEqual(false);

    const unfinishedProgress: Progress = { fullyLoadedFractionRanges: [{ start: 0, end: 0.5 }] };
    provider.extensionPoint.progressCallback(unfinishedProgress);
    expect(player._isPlaying).toEqual(false);

    const finishedProgress: Progress = { fullyLoadedFractionRanges: [{ start: 0, end: 1 }] };
    provider.extensionPoint.progressCallback(finishedProgress);
    expect(player._isPlaying).toEqual(true);
  });

  it("measures preloading performance", async () => {
    const provider = new TestProvider({ getMessages: async () => getMessagesResult });
    const client = new TestRunClient({ shouldLoadDataBeforePlaying: true });
    const player = new AutomatedRunPlayer(provider, client);
    let emitStateCalls = 0;
    player.setListener(async () => {
      emitStateCalls += 1;
    });

    expect(client.markPreloadStart.mock.calls.length).toBe(0);
    expect(client.markPreloadEnd.mock.calls.length).toBe(0);
    expect(emitStateCalls).toBe(0);

    player.setSubscriptions([{ topic: "/foo/bar", format: "parsedMessages" }]);
    await delay(AUTOMATED_RUN_START_DELAY + 10);

    // Preloading has started but not finished.
    expect(client.markPreloadStart.mock.calls.length).toBe(1);
    expect(client.markPreloadEnd.mock.calls.length).toBe(0);
    // _emitState called on initialization.
    expect(emitStateCalls).toBe(1);

    expect(player._initialized).toEqual(true);
    expect(player._isPlaying).toEqual(false);

    // Partially preloaded.
    provider.extensionPoint.progressCallback({ fullyLoadedFractionRanges: [{ start: 0, end: 0.5 }] });
    expect(player._isPlaying).toEqual(false);

    expect(client.markPreloadStart.mock.calls.length).toBe(1);
    expect(client.markPreloadEnd.mock.calls.length).toBe(0);
    // _emitState called on progress.
    expect(emitStateCalls).toBe(2);

    // Finish preloading.
    provider.extensionPoint.progressCallback({ fullyLoadedFractionRanges: [{ start: 0, end: 1 }] });
    await delay(0);
    // Finished preloading.
    expect(client.markPreloadStart.mock.calls.length).toBe(1);
    expect(client.markPreloadEnd.mock.calls.length).toBe(1);
    // _emitState called on progress (and also during playback.)
    expect(emitStateCalls).toBeGreaterThan(2);
    expect(player._isPlaying).toEqual(true);
  });

  it("makes calls to getMessages with the correct frames", async () => {
    const frames = [];
    const provider = new TestProvider({
      getMessages: async (startTime, endTime) => {
        frames.push({ startTime, endTime });
        return getMessagesResult;
      },
    });
    const client = new TestRunClient();

    const player = new AutomatedRunPlayer(provider, client);
    player.setSubscriptions([{ topic: "/foo/bar", format: "parsedMessages" }]);
    await delay(AUTOMATED_RUN_START_DELAY + 10);
    expect(player._initialized).toEqual(true);
    expect(player._isPlaying).toEqual(true);
    const listener = { signal: signal() };
    player.setListener(() => {
      return listener.signal;
    });
    while (!client.finished) {
      // Resolve the previous emit promise, then go to the next frame.
      const previousEmitSignal = listener.signal;
      listener.signal = signal();
      previousEmitSignal.resolve();
      await delay(1);
    }
    expect(frames).toMatchSnapshot();
  });

  it("makes calls to getMessages within the provider's range", async () => {
    const provider = new TestProvider({
      getMessages: async (startTime, endTime) => {
        expect(TimeUtil.compare(startTime, defaultStart) >= 0).toBeTruthy();
        expect(TimeUtil.compare(endTime, defaultEnd) <= 0).toBeTruthy();
        return getMessagesResult;
      },
    });
    const client = new TestRunClient();
    client.msPerFrame = 500;
    const player = new AutomatedRunPlayer(provider, client);
    player.setMessageOrder("headerStamp");
    player.setSubscriptions([{ topic: "/foo/bar", format: "parsedMessages" }]);
    const listener = { signal: signal() };
    player.setListener(() => listener.signal);
    while (!client.finished) {
      // Resolve the previous emit promise, then go to the next frame.
      const previousEmitSignal = listener.signal;
      listener.signal = signal();
      previousEmitSignal.resolve();
      await delay(1);
    }
  });

  async function setupEventLoopTest() {
    let getMessagesCallCount = 0;
    const getMessagesSignal = { signal: signal() };
    const provider = new TestProvider({
      getMessages: async () => {
        getMessagesCallCount++;
        return getMessagesSignal.signal;
      },
    });
    function resolveNextGetMessages() {
      const previousGetMessagesSignal = getMessagesSignal.signal;
      getMessagesSignal.signal = signal();
      previousGetMessagesSignal.resolve(getMessagesResult);
    }

    const listener = { signal: signal() };
    function resolveNextEmitState() {
      const previousEmitSignal = listener.signal;
      listener.signal = signal();
      previousEmitSignal.resolve();
    }

    const client = new TestRunClient();
    const player = new AutomatedRunPlayer(provider, client);
    player.setSubscriptions([{ topic: "/foo/bar", format: "parsedMessages" }]);
    await delay(AUTOMATED_RUN_START_DELAY + 10);
    player.setListener(() => listener.signal);

    // We make one getMessages call at the beginning that we have to resolve.
    resolveNextGetMessages();
    // Get us into the event loop.
    while (!client.frameStarted) {
      resolveNextEmitState();
      await delay(10);
    }

    // Make sure we've resolved our last emit state, so that there is no state waiting to go out.
    resolveNextEmitState();
    await delay(10);

    // Now we're in the run loop.
    // Reset the call count to 0.
    getMessagesCallCount = 0;

    return {
      player,
      client,
      resolveNextGetMessages,
      resolveNextEmitState,
      getGetMessagesCallCount: () => getMessagesCallCount,
    };
  }

  it("awaits the previous emit promise before calling getMessages", async () => {
    const { resolveNextGetMessages, resolveNextEmitState, getGetMessagesCallCount } = await setupEventLoopTest();

    resolveNextGetMessages();
    await delay(10);
    // Don't call getMessages until the previous listener has resolved.
    expect(getGetMessagesCallCount()).toEqual(0);
    resolveNextEmitState();
    await delay(10);
    expect(getGetMessagesCallCount()).toEqual(1);
  });
});
