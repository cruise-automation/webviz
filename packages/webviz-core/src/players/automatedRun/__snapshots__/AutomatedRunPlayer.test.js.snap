// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutomatedRunPlayer makes calls to getMessages with the correct frames 1`] = `
Array [
  Object {
    "endTime": Object {
      "nsec": 99000000,
      "sec": 10,
    },
    "startTime": Object {
      "nsec": 0,
      "sec": 10,
    },
  },
  Object {
    "endTime": Object {
      "nsec": 99000000,
      "sec": 10,
    },
    "startTime": Object {
      "nsec": 99000000,
      "sec": 10,
    },
  },
  Object {
    "endTime": Object {
      "nsec": 99000000,
      "sec": 30,
    },
    "startTime": Object {
      "nsec": 800000000,
      "sec": 29,
    },
  },
  Object {
    "endTime": Object {
      "nsec": 99000000,
      "sec": 50,
    },
    "startTime": Object {
      "nsec": 800000000,
      "sec": 49,
    },
  },
  Object {
    "endTime": Object {
      "nsec": 99000000,
      "sec": 70,
    },
    "startTime": Object {
      "nsec": 800000000,
      "sec": 69,
    },
  },
  Object {
    "endTime": Object {
      "nsec": 99000000,
      "sec": 90,
    },
    "startTime": Object {
      "nsec": 800000000,
      "sec": 89,
    },
  },
]
`;
